src/pytorch_lightning/core/module.py

```diff
@@ -1478,16 +1476,14 @@ def toggle_optimizer(self, optimizer: Union[Optimizer, LightningOptimizer], opti
                 param.requires_grad = param_requires_grad_state[param]
         self._param_requires_grad_state = param_requires_grad_state
 
-    def untoggle_optimizer(self, optimizer_idx: int) -> None:
+    def untoggle_optimizer(self, optimizer: Union[Optimizer, LightningOptimizer]) -> None:
         """Resets the state of required gradients that were toggled with :meth:`toggle_optimizer`.
 
-        This is only called automatically when automatic optimization is enabled and multiple optimizers are used.
-
         Args:
-            optimizer_idx: The index of the optimizer to untoggle.
+            optimizer: The optimizer to untoggle.
         """
-        for opt_idx, opt in enumerate(self.trainer.optimizers):
-            if optimizer_idx != opt_idx:
+        for opt in self.trainer.optimizers:
+            if not (opt == optimizer or (isinstance(optimizer, LightningOptimizer) and opt == optimizer.optimizer)):
```

Comment:
nit: I think this should check by reference
```suggestion
            if not (opt is optimizer or (isinstance(optimizer, LightningOptimizer) and opt is optimizer.optimizer)):
```